django/db/models/functions/comparison.py

```diff
@@ -175,24 +177,31 @@ def join(self, args):
         )
 
     def as_postgresql(self, compiler, connection, **extra_context):
-        if (
-            not connection.features.is_postgresql_16
-            or connection.features.uses_server_side_binding
-        ):
-            copy = self.copy()
-            copy.set_source_expressions(
-                [
-                    Cast(expression, TextField()) if index % 2 == 0 else expression
-                    for index, expression in enumerate(copy.get_source_expressions())
-                ]
-            )
-            return super(JSONObject, copy).as_sql(
-                compiler,
-                connection,
-                function="JSONB_BUILD_OBJECT",
-                **extra_context,
+        # It is required to cast keys to text while using JSONB_BUILD_OBJECT
+        # and also while using JSON_OBJECT in PostgreSQL 16+ with server-side
+        # binding.
+        # It is not technically required to cast keys to text while using
+        # JSON_OBJECT in PostgreSQL 16+ without server-side binding, but it is
+        # done for consistency.
+        copy = self.copy()
+        copy.set_source_expressions(
+            [
+                Cast(expression, TextField()) if index % 2 == 0 else expression
+                for index, expression in enumerate(copy.get_source_expressions())
+            ]
+        )
+
+        if connection.features.is_postgresql_16:
+            return copy.as_native(
+                compiler, connection, returning="JSONB", **extra_context
             )
```

Comment:
Just to share some of my thoughts when reviewing, I was wondering if we could do anything to isolate the wrapping of keys in parentheses to the Postgres implementation, something like
```suggestion
            class ArgJoiner:
                def join(self, args):
                    pairs = zip(args[::2], args[1::2], strict=True)
                    # Wrap 'key' in parentheses due to postgres cast :: shortcut syntax.
                    return ", ".join(f"({key}) VALUE {value}" for key, value in pairs)
            return copy.as_native(
                compiler, connection, returning="JSONB", arg_joiner=ArgJoiner(), **extra_context
            )
```
with `as_native` being something like 
```python
    def as_native(self, compiler, connection, *, returning, arg_joiner=None, **extra_context):
        class ArgJoiner:
            def join(self, args):
                pairs = zip(args[::2], args[1::2], strict=True)
                return ", ".join([" VALUE ".join(pair) for pair in pairs])
        return self.as_sql(
            compiler,
            connection,
            arg_joiner=arg_joiner or ArgJoiner(),
            template=f"%(function)s(%(expressions)s RETURNING {returning})",
            **extra_context,
        )
```

But I think it's a lot more complicated for the value add given there are no known issues in wrapping keys in parenthesis for oracle 
So I don't think that would be worth doing :+1: